### PR TITLE
compare consts using ptr equality

### PR DESCRIPTION
**Pull request by lcnr**
Similar to `Ty<'tcx>` this PR now compares `Const`s using pointer equality.

To prevent ourselves from accidentally creating a const on the stack I added a new
zero sized field to `Const`. This field has to be `public`, to allow pattern matching on consts
and has a private field, to prevent construction outside of `rustc_middle::ty`.

Let's see if this actually changes perf in any meaningful way :shrug:
I did not yet update the test output.

### CODE REVIEW

File: src/librustc_middle/ty/sty.rs [hunk 1]
```diff
@@ -2178,21 +2180,38 @@ impl<'tcx> TyS<'tcx> {
...
 pub struct Const<'tcx> {
     pub ty: Ty<'tcx>,
 
     pub val: ConstKind<'tcx>,
+
+    /// We compare `Const` by pointer equality. This would be incorrect
+    /// if we were able to create Const values on the stack.
+    ///
+    /// Only allow the creation of consts using `tcx.mk_const(ty, val)`.
+    pub _priv: PrivateMarker,
```
Comment (bjorn3) [inline src/librustc_middle/ty/sty.rs:2194]: Why is this `pub`?

Comment (lcnr) [reply src/librustc_middle/ty/sty.rs:2194]: > This field has to be public, to allow pattern matching on consts

Not completely sure if this is the best approach :thinking:

Comment (iwikal) [reply src/librustc_middle/ty/sty.rs:2194]: Private fields don't prevent pattern matching, you can still do this:
```rust
Const {
  ty,
  val,
  ..
}
```

Comment (lcnr) [reply src/librustc_middle/ty/sty.rs:2194]: I thought I tried that and remembered that I had a similar problem with `#[non_exhaustive]` in the past :sweat_smile: 

Well, looks like I should clean this up then.
Thanks x.x

File: src/librustc_middle/ty/sty.rs [hunk 2]
```diff
@@ -2178,21 +2180,38 @@ impl<'tcx> TyS<'tcx> {
 }
 
 /// Typed constant value.
-#[derive(Copy, Clone, Debug, Hash, RustcEncodable, RustcDecodable, Eq, PartialEq, Ord, PartialOrd)]
+#[derive(Debug, Hash, RustcEncodable, RustcDecodable, Eq, Ord, PartialOrd)]
```
Comment (matthewjasper) [inline src/librustc_middle/ty/sty.rs:2183]: Hash can hash the address now.

**Comment by lcnr** [comment]
I am currently getting the following errors:
```
thread 'rustc' panicked at 'forcing query with already existing `DepNode`
- query-key: Canonical { max_universe: U0, variables: [], value: ParamEnvAnd { param_env: ParamEnv { caller_bounds: [Binder(TraitPredicate(<[T; N] as array::LengthAtMost32>)), Binder(TraitPredicate(<T as marker::Sized>))], reveal: UserFacing, def_id: None }, value: array::iter::IntoIter<T, N> } }
- dep-node: implied_outlives_bounds(4f246e06f2621b4f-6e0eacf16f43cb9)', /home/programming/rust6/src/libstd/macros.rs:16:9
```

Even after looking at #48923 I am still kind of stuck with what to do here.

File: src/librustc_middle/ty/context.rs [hunk 3]
```diff
@@ -1974,9 +1986,9 @@ impl<'tcx> Borrow<[Predicate<'tcx>]> for Interned<'tcx, List<Predicate<'tcx>>> {
     }
 }
 
-impl<'tcx> Borrow<Const<'tcx>> for Interned<'tcx, Const<'tcx>> {
-    fn borrow<'a>(&'a self) -> &'a Const<'tcx> {
-        &self.0
+impl<'tcx> Borrow<super::sty::InternedConst<'tcx>> for Interned<'tcx, Const<'tcx>> {
+    fn borrow<'a>(&'a self) -> &'a super::sty::InternedConst<'tcx> {
+        unsafe { mem::transmute(self) }
```
Comment (bjorn3) [inline src/librustc_middle/ty/context.rs:1991]: `Const` and `InternedConst` may have a different layout.

Comment (bjorn3) [inline src/librustc_middle/ty/context.rs:1991]: `Const` and `InternedConst` may have a different layout.

Comment (lcnr) [reply src/librustc_middle/ty/context.rs:1991]: I thought it would be fine as they are both `repr(C)`.

Comment (bjorn3) [reply src/librustc_middle/ty/context.rs:1991]: Missed the `#[repr(C)]`.

**Comment by lcnr** [comment]
Will close this PR for now as I don't know how to fix the current issues and we do not often compare `Const`s, so I don't know if this is worth the effort.

**Pull request closed without merging**